src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,223 @@
+  const std::vector<std::string> args() const { return args_; }
+  Sign sign() { return sign_; }
+
+ private:
+  std::string name_;
+  std::vector<std::string> args_;
+  Sign sign_;
+};
+
+// Attribute corresponds to an attribute of a principal with the form
+// <Principal> <pred_name>(arg_1, ..., arg_n)
+class Attribute {
+ public:
+  explicit Attribute(Principal principal, Predicate predicate)
+    : principal_(principal), predicate_(predicate) {}
+  const Principal principal() const { return principal_; }
+  const Predicate predicate() const { return predicate_; }
+ private:
+  Principal principal_;
+  Predicate predicate_;
+};
+
+// CanActAs corresponds to a canActAs expression of the form
+// <principal_1> canActAs <principal_2>
+class CanActAs {
+ public:
+  explicit CanActAs(Principal left_principal, Principal right_principal)
+    : left_principal_(left_principal),
+      right_principal_(right_principal) {}
+  const Principal left_principal() const { return left_principal_; }
+  const Principal right_principal() const { return right_principal_; }
+
+ private:
+   Principal left_principal_;
+   Principal right_principal_;
+};
+
+// BaseFact corresponds to a base-case fact that cannot recursively include other 
+// facts through "canSay". "Fact" includes the recursive case. It is necessary 
+// that BaseFact is a separate class from Fact because the RHS of conditional 
+// assertions can only contain BaseFacts.
+// Base facts have three forms:
+//  - A predicate
+//  - An attribute
+//  - A canActAs expression
+// The particular choice of form is represented with a variant which is 
+// encapsulated.
+class BaseFact {
+ public:
+  explicit BaseFact(Predicate predicate) : value_(std::move(predicate)) {};
+  explicit BaseFact(Attribute attribute) : value_(std::move(attribute)) {};
+  explicit BaseFact(CanActAs canActAs) : value_(std::move(canActAs)) {};
+  const std::variant<Predicate, Attribute, CanActAs> value() const {
+    return value_;
+  }
+
+ private:
+  std::variant<Predicate, Attribute, CanActAs> value_;
+};
+
+// CanSay and Fact are forward-declared because they are mutually recursive.
+class CanSay;
+class Fact;
+
+// CanSay corresponds to an expression of the form <principal> canSay Fact
+class CanSay {
+  public:
+    explicit CanSay(Principal principal, std::shared_ptr<Fact>& fact)
+      : principal_(principal), fact_(std::move(fact)) {}
+    const Principal principal() const { return principal_; }
+    const std::shared_ptr<Fact> fact() const { return fact_; }
+  private:
+    Principal principal_;
+    std::shared_ptr<Fact> fact_;
+};
+
+// Fact corresponds to either a base fact or a an expression of the form 
+// <principal> canSay 
+class Fact {
+ public:
+  explicit Fact(BaseFact base_fact) : value_(std::move(base_fact)) {}
+  explicit Fact(std::shared_ptr<CanSay>& can_say)
+    : value_(std::move(can_say)) {}
+  const std::variant<BaseFact, std::shared_ptr<CanSay>> value() const {
```

Comment:
done

Reply:
> The high level theme of most of my comments here: I was distracted by the accessors making copies, but I do wonder whether they should be returning their child nodes outside of the class at all. Generally, an AST tries to make its possible children a bit of a private matter and expects to receive a Visitor class, which can handle any given child. This is particularly nice for operations in which AST nodes should mostly be considered separately anyway. I think most of the accessors should probably go and instead a `visit` method should be used to consider the AST contents.

I was thinking about making a visitor as well, but a visitor in C++ didn't seem all that useful because since it lacks generics, the only way to write a visitor with a return is to have the return type `void *`. The other option is to never return anything with a visitor and merely fake returns by adding field to the visitor and side-effecting in it, but I find the side-effect style confusing. So I didn't like any of the visitors in C++ that I saw. Also, the functions that need to be implemented on the AST (translation into a lower level datalog IR, and just printing) need to do something in between visits to the children, so at least the style of visitor that keeps the order of traversal fixed for the AST does not work here.

Reply:
Maybe the style of visitor that: 1) can return `void *`, 2) and does not fix the order of traversal in children would be fine.